File: packages/types/src/external-systems.ts
/**
 * Dependency-registry types — the `external_systems` table populated by the
 * manifest parsers during ingestion. Mirrors the Pydantic models served by
 * `/api/repos/{id}/external-systems`.
 */

/**
 * Canonical I/O-boundary kinds. Mirrors `IO_KINDS` in the Python classifier
 * (`packages/core/.../ingestion/external_systems/io_kind.py`). The cross-
 * language parity guard lives in `__tests__/contracts.test.ts` (TS half) and
 * `tests/unit/ingestion/test_io_kind.py` (Python half). Change one, change
 * all three.
 */
export const C4_IO_KINDS = [
  "db",
  "network",
  "filesystem",
  "subprocess",
  "lock",
] as const;

/** A dependency's I/O-boundary type, or null when it isn't typed. */
export type C4IoKind = (typeof C4_IO_KINDS)[number];

/** One declared third-party dependency. */
export interface ExternalSystemEntry {
  name: string;
  display_name: string;
  /** npm | pypi | cargo | gomod | nuget | ... */
  ecosystem: string;
  /** framework | service | tool | library */
  category: string;
  /** db | network | filesystem | subprocess | lock, or null when untyped. */
  io_kind: C4IoKind | null;
  version: string | null;
  /** Manifest path the dependency was declared in, e.g. "packages/web/package.json". */
  declared_in: string;
  is_dev_dep: boolean;
}

/** The full dependency registry for a repository. */
export interface ExternalSystemsRegistry {
  items: ExternalSystemEntry[];
  total: number;
  prod_count: number;
  dev_count: number;
  ecosystems: string[];
  manifests: string[];
}

export type ExternalSystemLinkState = "linked" | "unlinked";
export type ExternalSystemsSummaryScope = "primary" | "all";

/** One canonical package with declaration and persisted graph-usage aggregates. */
export interface ExternalSystemSummaryEntry {
  package_key: string;
  name: string;
  display_name: string;
  ecosystem: string;
  category: string;
  io_kind: C4IoKind | null;
  runtime_declared: boolean;
  dev_declared: boolean;
  declaration_count: number;
  manifest_count: number;
  versions: string[];
  versions_total: number;
  versions_truncated: boolean;
  multiple_versions: boolean;
  external_node_count: number;
  import_edge_count: number;
  importing_file_count: number;
  link_state: ExternalSystemLinkState;
}

/** Bounded package summaries for the external-dependency scan surface. */
export interface ExternalSystemsSummary {
  items: ExternalSystemSummaryEntry[];
  returned: number;
  total_packages: number;
  limit: number;
  offset: number;
  truncated: boolean;
  scope: ExternalSystemsSummaryScope;
  excluded_declarations: number;
  total_declarations: number;
  runtime_packages: number;
  dev_only_packages: number;
  observed_packages: number;
  linked_packages: number;
  unlinked_packages: number;
  linked_without_imports: number;
  ecosystems: string[];
  manifest_count: number;
}

export type ExternalSystemMatchBasis = "exact" | "subpath" | "mapped" | "mixed" | "unresolved";

export interface ExternalSystemGraphTarget {
  node_id: string;
  match_basis: Exclude<ExternalSystemMatchBasis, "mixed" | "unresolved">;
}

export interface ExternalSystemRelationshipNode {
  aggregate_key: string;
  label: string;
  community_id: number;
  importing_file_count: number;
  import_edge_count: number;
  top_file: string | null;
}

export interface ExternalSystemRelationshipEdge {
  source: string;
  target: string;
  import_edge_count: number;
}

export interface ExternalSystemRelationshipGraph {
  package_key: string;
  package_name: string;
  package_node_id: string;
  match_basis: ExternalSystemMatchBasis;
  matched_external_nodes: ExternalSystemGraphTarget[];
  matched_external_nodes_total: number;
  matched_external_nodes_truncated: boolean;
  evidence_target_limit: number;
  evidence_truncated: boolean;
  nodes: ExternalSystemRelationshipNode[];
  edges: ExternalSystemRelationshipEdge[];
  aggregate_total: number;
  aggregate_returned: number;
  edge_total: number;
  edge_returned: number;
  importing_file_total: number;
  import_edge_total: number;
  node_limit: number;
  edge_limit: number;
  truncated: boolean;
  scope: ExternalSystemsSummaryScope;
}

export interface ExternalSystemImportingFile {
  path: string;
  language: string;
  import_edge_count: number;
  matched_external_node_count: number;
}

export interface ExternalSystemImportingFiles {
  package_key: string;
  aggregate_key: string;
  items: ExternalSystemImportingFile[];
  total: number;
  returned: number;
  limit: number;
  offset: number;
  truncated: boolean;
  scope: ExternalSystemsSummaryScope;
}

File: packages/server/src/repowise/server/schemas/external_systems.py
"""Dependency-registry response models (``external_systems`` table)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class ExternalSystemEntry(BaseModel):
    """One declared third-party dependency."""

    name: str
    display_name: str
    ecosystem: str  # npm | pypi | cargo | gomod | nuget | ...
    category: str  # framework | service | tool | library
    io_kind: str | None = None  # db | network | filesystem | subprocess | lock | null
    version: str | None = None
    declared_in: str  # manifest path, e.g. "packages/web/package.json"
    is_dev_dep: bool = False


class ExternalSystemsResponse(BaseModel):
    """The full dependency registry for a repository."""

    items: list[ExternalSystemEntry]
    total: int
    prod_count: int
    dev_count: int
    ecosystems: list[str]
    manifests: list[str]


ExternalSystemLinkState = Literal["linked", "unlinked"]
ExternalSystemsSummaryScope = Literal["primary", "all"]
ExternalSystemMatchBasis = Literal["exact", "subpath", "mapped", "mixed", "unresolved"]


class ExternalSystemSummaryEntry(BaseModel):
    """One canonical package with declaration and graph-usage aggregates."""

    package_key: str
    name: str
    display_name: str
    ecosystem: str
    category: str
    io_kind: str | None = None
    runtime_declared: bool
    dev_declared: bool
    declaration_count: int
    manifest_count: int
    versions: list[str]
    versions_total: int
    versions_truncated: bool
    multiple_versions: bool
    external_node_count: int
    import_edge_count: int
    importing_file_count: int
    link_state: ExternalSystemLinkState


class ExternalSystemsSummaryResponse(BaseModel):
    """Bounded package summaries for the external-dependency scan surface."""

    items: list[ExternalSystemSummaryEntry]
    returned: int
    total_packages: int
    limit: int
    offset: int
    truncated: bool
    scope: ExternalSystemsSummaryScope
    excluded_declarations: int
    total_declarations: int
    runtime_packages: int
    dev_only_packages: int
    observed_packages: int
    linked_packages: int
    unlinked_packages: int
    linked_without_imports: int
    ecosystems: list[str]
    manifest_count: int


class ExternalSystemGraphTarget(BaseModel):
    """One persisted external graph node linked to the selected package."""

    node_id: str
    match_basis: Literal["exact", "subpath", "mapped"]


class ExternalSystemRelationshipNode(BaseModel):
    """A first-party graph community that imports the selected package."""

    aggregate_key: str
    label: str
    community_id: int
    importing_file_count: int
    import_edge_count: int
    top_file: str | None = None


class ExternalSystemRelationshipEdge(BaseModel):
    source: str
    target: str
    import_edge_count: int


class ExternalSystemRelationshipGraphResponse(BaseModel):
    """Bounded aggregate-first relationship graph for one declared package."""

    package_key: str
    package_name: str
    package_node_id: str
    match_basis: ExternalSystemMatchBasis
    matched_external_nodes: list[ExternalSystemGraphTarget]
    matched_external_nodes_total: int
    matched_external_nodes_truncated: bool
    evidence_target_limit: int
    evidence_truncated: bool
    nodes: list[ExternalSystemRelationshipNode]
    edges: list[ExternalSystemRelationshipEdge]
    aggregate_total: int
    aggregate_returned: int
    edge_total: int
    edge_returned: int
    importing_file_total: int
    import_edge_total: int
    node_limit: int
    edge_limit: int
    truncated: bool
    scope: ExternalSystemsSummaryScope


class ExternalSystemImportingFile(BaseModel):
    path: str
    language: str
    import_edge_count: int
    matched_external_node_count: int


class ExternalSystemImportingFilesResponse(BaseModel):
    """One independently bounded page of files behind an aggregate node."""

    package_key: str
    aggregate_key: str
    items: list[ExternalSystemImportingFile]
    total: int
    returned: int
    limit: int
    offset: int
    truncated: bool
    scope: ExternalSystemsSummaryScope
